File: ReadMem.py
import subprocess
import importlib
import sys

required_modules = ['pymem']

def install_missing_modules(modules):
    try:
        pip = 'pip'
        importlib.import_module(pip)
    except ImportError:
        print(f"{pip} is not installed. Installing...")
        subprocess.check_call([sys.executable, "-m", "ensurepip", "--upgrade"])
    for module in modules:
        try:
            importlib.import_module(module)
        except ImportError:
            print(f"{module} is not installed. Installing...")
            subprocess.check_call([sys.executable, "-m", "pip", "install", module])

install_missing_modules(required_modules)

import pymem
import pymem.process
import os
 '''
Commented out code is from past versions of the script.
 '''
def read_board(pm, base_address, start_offset, row_count=30, col_count=30, row_stride=0x20):
    board = []
    for row in range(row_count):
        row_offset = start_offset + (row * row_stride)
        row_bytes = pm.read_bytes(base_address + row_offset, col_count)
        #row_symbols = ['X' if byte == 143 else 'O' for byte in row_bytes]
        row_symbols = ['\033[91mX\033[0m' if byte == 143 else 'O' for byte in row_bytes]
        #board.append([byte for byte in row_bytes])
        board.append(row_symbols)
    return board

def main():
    os.system('cls')
    process_name = "winmine.exe"
    base_offset = 0x5361  # Offset of the first cell (top-left)
    width_offset = 0x5334
    height_offset = 0x5338

    try:
        pm = pymem.Pymem(process_name)
        module_base = pymem.process.module_from_name(pm.process_handle, process_name).lpBaseOfDll

        # Read width and height dynamically from process memory
        col_count = pm.read_int(module_base + width_offset)
        row_count = pm.read_int(module_base + height_offset)

        board = read_board(pm, module_base, base_offset, row_count=row_count, col_count=col_count)

        os.system('cls')
        for row in board:
            #print(row)
            print(' '.join(row))

    except pymem.exception.ProcessNotFound:
        print(f"Error: Process '{process_name}' not found.")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

    input("Press Enter to exit...")
    os.system('cls')

if __name__ == "__main__":
    main()

File: ReadMem.c
#include <windows.h>
#include <tlhelp32.h>
#include <psapi.h>
#include <stdio.h>

#define PROCESS_NAME "winmine.exe"
// Build: cl winmine_reader.c /link psapi.lib
// Offsets for classic Minesweeper
#define BASE_OFFSET  0x5361
#define WIDTH_OFFSET 0x5334
#define HEIGHT_OFFSET 0x5338
#define ROW_STRIDE   0x20

DWORD FindProcessId(const char* processName) {
    PROCESSENTRY32 processEntry;
    processEntry.dwSize = sizeof(PROCESSENTRY32);

    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshot == INVALID_HANDLE_VALUE)
        return 0;

    if (Process32First(snapshot, &processEntry)) {
        do {
            if (_stricmp(processEntry.szExeFile, processName) == 0) {
                DWORD pid = processEntry.th32ProcessID;
                CloseHandle(snapshot);
                return pid;
            }
        } while (Process32Next(snapshot, &processEntry));
    }

    CloseHandle(snapshot);
    return 0;
}

void SetConsoleColor(WORD color) {
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    SetConsoleTextAttribute(hConsole, color);
}

void ReadBoard(HANDLE hProcess, uintptr_t baseAddress, int startOffset, int rowCount, int colCount) {
    BYTE buffer[32];
    for (int row = 0; row < rowCount; row++) {
        int rowOffset = startOffset + (row * ROW_STRIDE);
        SIZE_T bytesRead;
        if (ReadProcessMemory(hProcess, (LPCVOID)(baseAddress + rowOffset), buffer, colCount, &bytesRead)) {
            for (int col = 0; col < colCount; col++) {
                if (buffer[col] == 0x8F) {
                    SetConsoleColor(FOREGROUND_RED | FOREGROUND_INTENSITY);
                    printf("X ");
                    SetConsoleColor(FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE); // reset to white
                } else {
                    printf("O ");
                }
            }
            printf("\n");
        } else {
            printf("Failed to read row %d\n", row);
        }
    }
}

int main() {
    DWORD pid = FindProcessId(PROCESS_NAME);
    if (pid == 0) {
        printf("Error: Process '%s' not found.\n", PROCESS_NAME);
        return 1;
    }

    HANDLE hProcess = OpenProcess(PROCESS_VM_READ | PROCESS_QUERY_INFORMATION, FALSE, pid);
    if (!hProcess) {
        printf("Error: Failed to open process.\n");
        return 1;
    }

    HMODULE hMods[1024];
    DWORD cbNeeded;
    if (!EnumProcessModules(hProcess, hMods, sizeof(hMods), &cbNeeded)) {
        printf("Error: Could not enumerate process modules.\n");
        CloseHandle(hProcess);
        return 1;
    }

    uintptr_t baseAddress = (uintptr_t)hMods[0];
    int width = 0, height = 0;
    SIZE_T bytesRead;

    ReadProcessMemory(hProcess, (LPCVOID)(baseAddress + WIDTH_OFFSET), &width, sizeof(int), &bytesRead);
    ReadProcessMemory(hProcess, (LPCVOID)(baseAddress + HEIGHT_OFFSET), &height, sizeof(int), &bytesRead);

    ReadBoard(hProcess, baseAddress, BASE_OFFSET, height, width);

    CloseHandle(hProcess);

    printf("Press Enter to exit...");
    getchar();
    return 0;
}